### ISSUE TITLE
Race condition between ProcessQueues and SubscribeAsync in BrokerService

### ISSUE THREAD

**Issue by BoeseB** [loekd/ServiceFabric.PubSubActors#90]
We noticed some strange InvalidOperation exceptions in the past.

This was due to new services subscribing while the broker is already running and before this action completed the process loop got executed. In this state the new queue for the subscription was already added to the `_subscriptions ` dictionary but the transaction for creating the queue was not completed which leads to an error when trying to execute any operation on that new collection like the `GetCountAsync` call in the `ProcessQueues` method. 

The way we currently solved it is to call `_subscriptionFactory.CreateAsync(subscriptionDetails)` in the SubscribeAsync method instead of the overload which accepts the transaction.
We see no problem in having a stray dictionary if the code fails after the creation of this dictionary and preferred this over some locks between the processing loop and the `SubscribeAsync `method.

```
System.InvalidOperationException:
   at Microsoft.ServiceFabric.Data.Collections.DistributedQueue`1.ThrowIfNotRegistered (Microsoft.ServiceFabric.Data.Impl, Version=6.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35)
   at Microsoft.ServiceFabric.Data.Collections.DistributedQueue`1+<GetCountAsync>d__65.MoveNext (Microsoft.ServiceFabric.Data.Impl, Version=6.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at SoCreate.ServiceFabric.PubSub.Helpers.TimeoutRetryHelper+<>c__DisplayClass3_0`1+<<ExecuteInTransaction>b__0>d.MoveNext (SoCreate.ServiceFabric.PubSub, Version=11.0.0.0, Culture=neutral, PublicKeyToken=nullSoCreate.ServiceFabric.PubSub, Version=11.0.0.0, Culture=neutral, PublicKeyToken=null: service-fabric-pub-sub\src\SoCreate.ServiceFabric.PubSub\Helpers\TimeoutRetryHelper.cs SoCreate.ServiceFabric.PubSub, Version=11.0.0.0, Culture=neutral, PublicKeyToken=null: 48)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at SoCreate.ServiceFabric.PubSub.Helpers.TimeoutRetryHelper+<Execute>d__5`1.MoveNext (SoCreate.ServiceFabric.PubSub, Version=11.0.0.0, Culture=neutral, PublicKeyToken=nullSoCreate.ServiceFabric.PubSub, Version=11.0.0.0, Culture=neutral, PublicKeyToken=null: service-fabric-pub-sub\src\SoCreate.ServiceFabric.PubSub\Helpers\TimeoutRetryHelper.cs SoCreate.ServiceFabric.PubSub, Version=11.0.0.0, Culture=neutral, PublicKeyToken=null: 134)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at SoCreate.ServiceFabric.PubSub.Helpers.TimeoutRetryHelper+<ExecuteInTransaction>d__3`1.MoveNext (SoCreate.ServiceFabric.PubSub, Version=11.0.0.0, Culture=neutral, PublicKeyToken=nullSoCreate.ServiceFabric.PubSub, Version=11.0.0.0, Culture=neutral, PublicKeyToken=null: service-fabric-pub-sub\src\SoCreate.ServiceFabric.PubSub\Helpers\TimeoutRetryHelper.cs SoCreate.ServiceFabric.PubSub, Version=11.0.0.0, Culture=neutral, PublicKeyToken=null: 60)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at SoCreate.ServiceFabric.PubSub.State.SubscriptionOrdered+<GetQueueCount>d__13.MoveNext (SoCreate.ServiceFabric.PubSub, Version=11.0.0.0, Culture=neutral, PublicKeyToken=nullSoCreate.ServiceFabric.PubSub, Version=11.0.0.0, Culture=neutral, PublicKeyToken=null: service-fabric-pub-sub\src\SoCreate.ServiceFabric.PubSub\State\SubscriptionOrdered.cs SoCreate.ServiceFabric.PubSub, Version=11.0.0.0, Culture=neutral, PublicKeyToken=null: 48)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at SoCreate.ServiceFabric.PubSub.BrokerService+<ProcessQueues>d__47.MoveNext (SoCreate.ServiceFabric.PubSub, Version=11.0.0.0, Culture=neutral, PublicKeyToken=nullSoCreate.ServiceFabric.PubSub, Version=11.0.0.0, Culture=neutral, PublicKeyToken=null: service-fabric-pub-sub\src\SoCreate.ServiceFabric.PubSub\BrokerService.cs SoCreate.ServiceFabric.PubSub, Version=11.0.0.0, Culture=neutral, PublicKeyToken=null: 462)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at SoCreate.ServiceFabric.PubSub.BrokerService+<ProcessAllQueuesAsync>d__43.MoveNext (SoCreate.ServiceFabric.PubSub, Version=11.0.0.0, Culture=neutral, PublicKeyToken=nullSoCreate.ServiceFabric.PubSub, Version=11.0.0.0, Culture=neutral, PublicKeyToken=null: service-fabric-pub-sub\src\SoCreate.ServiceFabric.PubSub\BrokerService.cs SoCreate.ServiceFabric.PubSub, Version=11.0.0.0, Culture=neutral, PublicKeyToken=null: 382)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at SoCreate.ServiceFabric.PubSub.BrokerService+<RunAsync>d__42.MoveNext (SoCreate.ServiceFabric.PubSub, Version=11.0.0.0, Culture=neutral, PublicKeyToken=nullSoCreate.ServiceFabric.PubSub, Version=11.0.0.0, Culture=neutral, PublicKeyToken=null: service-fabric-pub-sub\src\SoCreate.ServiceFabric.PubSub\BrokerService.cs SoCreate.ServiceFabric.PubSub, Version=11.0.0.0, Culture=neutral, PublicKeyToken=null: 350)
```

I hope this info helps you to improve this library.

**Comment by danadesrosiers** [comment]
Thanks for reporting this and taking the time to track down the root cause!  

I'm not sure I'm comfortable with letting it break and accepting that there will be some orphaned queues from time to time.  It seems like there could be some pitfalls that are not immediately clear.  I like the lock idea better.  I don't think that would be so bad because the subscribe won't take a long time and it's a relatively rare event, so it shouldn't have a noticeable impact.

A couple other possibilities:
1. We could wait until after the transaction is completed to add the new subscription to the _subscriptions dictionary.  This is unlikely to fail and if it does, we can do some clean up without affecting the main ProcessQueues loop.
2. We could more gracefully handle the case where the queue does not exist yet.  Maybe catch exceptions and log a warning and return `0` for `GetQueueCount()`.
3. We could put a flag on `ISubscription` that indicates it's status.  Set the flag to "pending" when creating the subscription, then "ready" after the transaction is committed.  Then in `ProcessAllQueues()`, filter out all subscriptions that aren't ready.

**Comment by BoeseB** [comment]
I know what you mean by the orphaned queues i was also not happy with it.
I also tried to replace the ConcurrentDictionary with an reliable one so it could be filled in the same transaction as the subscription queue is created and only once committed be picked up by the processing loop.
I went down a rabbit hole with this approach as i did not quite understand the broker stats stuff etc. 

I would prefer a transactional subscription over a lock between the two methods.

We also sometimes observe dropped subscriptions without calling unsubscribe, but have no deeper insights were this problem comes from. We guess it happens during swaps of the services from one node to another because of this local and reliable state mix in the broker service.
